File: code/postgresql-aurora/csharp/L04.cs
// Lesson 4: PostgreSQL from C# with Npgsql, on the course model (sql/schema.sql)
using System.Diagnostics;
using Npgsql;
using NpgsqlTypes;

namespace Learn.Pg;

static class L04
{
    public static async Task Connect()
    {
        // One NpgsqlDataSource per database for the whole application: it owns the connection pool
        await using var dataSource = Db.DataSource();
        await using var command = dataSource.CreateCommand(
            "SELECT current_user, current_setting('server_version'), (SELECT count(*) FROM ci.runs)");
        await using var reader = await command.ExecuteReaderAsync();
        await reader.ReadAsync();
        Console.WriteLine($"user: {reader.GetString(0)}");
        Console.WriteLine($"server: {reader.GetString(1)}");
        Console.WriteLine($"runs: {reader.GetInt64(2)}");
    }

    public static async Task Parameters()
    {
        await using var dataSource = Db.DataSource();

        // Positional parameters: $1, $2 are what PostgreSQL itself understands
        await using (var command = dataSource.CreateCommand("""
            SELECT run_id, started_at FROM ci.runs
            WHERE workflow_name = $1 AND started_at >= $2
            ORDER BY started_at, run_id
            LIMIT 2
            """))
        {
            command.Parameters.Add(new() { Value = "Rust course examples" });
            command.Parameters.Add(new() { Value = new DateTime(2026, 9, 13, 18, 0, 0, DateTimeKind.Utc) });
            await using var reader = await command.ExecuteReaderAsync();
            while (await reader.ReadAsync())
            {
                var startedAt = reader.GetDateTime(1);
                Console.WriteLine($"{reader.GetInt64(0)} {startedAt:O} Kind={startedAt.Kind}");
            }
        }

        // Named parameters: Npgsql rewrites @os into $1 before sending the query
        await using (var command = dataSource.CreateCommand("""
            SELECT count(*) FROM ci.jobs WHERE labels[1] = @os AND duration > @longer_than
            """))
        {
            command.Parameters.AddWithValue("os", "windows-latest");
            command.Parameters.AddWithValue("longer_than", TimeSpan.FromMinutes(2));
            Console.WriteLine($"windows jobs over 2 minutes: {await command.ExecuteScalarAsync()}");
        }

        // An array parameter replaces a list of values: = ANY ($1), not IN (...)
        await using (var command = dataSource.CreateCommand("""
            SELECT labels[1], count(*) FROM ci.jobs WHERE labels[1] = ANY ($1) GROUP BY 1 ORDER BY 1
            """))
        {
            command.Parameters.Add(new() { Value = new[] { "macos-latest", "windows-latest" } });
            await using var reader = await command.ExecuteReaderAsync();
            while (await reader.ReadAsync())
            {
                Console.WriteLine($"{reader.GetString(0)}: {reader.GetInt64(1)}");
            }
        }

        // DateTime.Kind decides the parameter's type: Utc is sent as timestamptz, Unspecified and Local as timestamp,
        // which the server converts to timestamptz in the session's time zone
        foreach (var options in new[] { "Timezone=UTC", "Timezone=America/Toronto" })
        {
            await using var zoned = Db.DataSource(options);
            foreach (var kind in new[] { DateTimeKind.Utc, DateTimeKind.Unspecified })
            {
                await using var command = zoned.CreateCommand("SELECT count(*) FROM ci.runs WHERE started_at >= $1");
                command.Parameters.Add(new() { Value = new DateTime(2026, 9, 14, 0, 0, 0, kind) });
                Console.WriteLine($"{options}, Kind={kind}: {await command.ExecuteScalarAsync()} runs since 2026-09-14 00:00");
            }
        }
        await using (var command = dataSource.CreateCommand("SELECT count(*) FROM ci.runs WHERE started_at >= $1"))
        {
            var toronto = new DateTimeOffset(2026, 9, 14, 0, 0, 0, TimeSpan.FromHours(-4));
            command.Parameters.Add(new() { Value = toronto });
            try
            {
                await command.ExecuteScalarAsync();
            }
            catch (ArgumentException e)
            {
                Console.WriteLine($"{e.GetType().Name}: {e.Message}");
            }
            command.Parameters[0].Value = toronto.ToUniversalTime();
            Console.WriteLine($"DateTimeOffset in UTC: {await command.ExecuteScalarAsync()} runs since 2026-09-14 00:00 in Toronto");
        }

        // A value that doesn't match the column's type is PostgreSQL's error, not Npgsql's
        await using (var command = dataSource.CreateCommand("SELECT count(*) FROM ci.runs WHERE run_id = $1"))
        {
            command.Parameters.Add(new() { Value = "34852867099" });
            try
            {
                await command.ExecuteScalarAsync();
            }
            catch (PostgresException e)
            {
                Console.WriteLine($"{e.SqlState}: {e.MessageText}");
            }
        }
    }

    public static async Task Pool()
    {
        foreach (var pooling in new[] { true, false })
        {
            await using var dataSource = Db.DataSource($"Pooling={pooling}");
            var backends = new HashSet<int>();
            for (var i = 0; i < 20; i++)
            {
                await using var connection = await dataSource.OpenConnectionAsync();
                backends.Add(connection.ProcessID);
            }
            Console.WriteLine($"Pooling={pooling}: 20 opens, {backends.Count} server process(es)");
        }

        // Five connections held at the same time are five server processes
        await using (var dataSource = Db.DataSource("Application Name=learn-pool"))
        {
            var connections = new List<NpgsqlConnection>();
            for (var i = 0; i < 5; i++)
            {
                connections.Add(await dataSource.OpenConnectionAsync());
            }
            foreach (var connection in connections)
            {
                await connection.CloseAsync();
            }
            await using var command = dataSource.CreateCommand("""
                SELECT state, count(*) FROM pg_stat_activity
                WHERE application_name = 'learn-pool' GROUP BY state ORDER BY state
                """);
            await using var reader = await command.ExecuteReaderAsync();
            while (await reader.ReadAsync())
            {
                Console.WriteLine($"after closing: {reader.GetInt64(1)} {reader.GetString(0)}");
            }
        }
        await using (var dataSource = Db.DataSource())
        {
            await using var command = dataSource.CreateCommand(
                "SELECT count(*) FROM pg_stat_activity WHERE application_name = 'learn-pool'");
            Console.WriteLine($"after disposing the data source: {await command.ExecuteScalarAsync()}");
        }
    }

    public static async Task Prepare()
    {
        // Npgsql prepares a statement only when asked, or after Auto Prepare Min Usages executions (default 5)
        // when Max Auto Prepare is above 0
        await using var dataSource = Db.DataSource("Max Auto Prepare=10");
        await using var connection = await dataSource.OpenConnectionAsync();
        await using var count = new NpgsqlCommand(
            "SELECT count(*) FROM pg_prepared_statements WHERE statement ~ 'ci\\.jobs WHERE'", connection);
        for (var i = 1; i <= 6; i++)
        {
            await using var command = new NpgsqlCommand("SELECT count(*) FROM ci.jobs WHERE run_id = $1", connection);
            command.Parameters.Add(new() { Value = 34852867099L });
            await command.ExecuteScalarAsync();
            Console.WriteLine($"after {i} execution(s): {await count.ExecuteScalarAsync()} prepared statement(s)");
        }
    }

    static IEnumerable<string[]> PackageRefs() =>
        File.ReadLines("../ladybugdb/data/ga/package_refs.csv").Skip(1).Select(line => line.Split(','));

    public static async Task Copy()
    {
        await using var dataSource = Db.DataSource();
        await using var connection = await dataSource.OpenConnectionAsync();
        await using (var create = new NpgsqlCommand("""
            DROP TABLE IF EXISTS ga.package_refs_import;
            CREATE TABLE ga.package_refs_import (LIKE ga.package_refs INCLUDING ALL);
            """, connection))
        {
            await create.ExecuteNonQueryAsync();
        }

        // COPY ... FROM STDIN in binary format: the client sends the rows, typed, in one stream
        await using (var writer = await connection.BeginBinaryImportAsync(
            "COPY ga.package_refs_import (project, package, version) FROM STDIN (FORMAT binary)"))
        {
            foreach (var fields in PackageRefs())
            {
                await writer.StartRowAsync();
                await writer.WriteAsync(fields[0], NpgsqlDbType.Text);
                await writer.WriteAsync(fields[1], NpgsqlDbType.Text);
                await writer.WriteAsync(fields[2], NpgsqlDbType.Text);
            }
            Console.WriteLine($"copied: {await writer.CompleteAsync()} rows");
        }

        await using var compare = new NpgsqlCommand("""
            SELECT (SELECT count(*) FROM (TABLE ga.package_refs EXCEPT TABLE ga.package_refs_import) AS d),
                   (SELECT count(*) FROM (TABLE ga.package_refs_import EXCEPT TABLE ga.package_refs) AS d)
            """, connection);
        await using var reader = await compare.ExecuteReaderAsync();
        await reader.ReadAsync();
        Console.WriteLine($"rows only in the server-side load: {reader.GetInt64(0)}, only in the client copy: {reader.GetInt64(1)}");
    }

    public static async Task ExerciseBatch()
    {
        await using var dataSource = Db.DataSource();
        await using var connection = await dataSource.OpenConnectionAsync();
        await using (var create = new NpgsqlCommand("""
            DROP TABLE IF EXISTS ga.package_refs_batch;
            CREATE TABLE ga.package_refs_batch (LIKE ga.package_refs INCLUDING ALL);
            """, connection))
        {
            await create.ExecuteNonQueryAsync();
        }

        // One NpgsqlBatch: every INSERT goes to the server in one round trip, in one implicit transaction
        var rows = PackageRefs().ToList();
        rows.Add(rows[0]);
        await using var batch = new NpgsqlBatch(connection);
        foreach (var fields in rows)
        {
            var command = new NpgsqlBatchCommand("INSERT INTO ga.package_refs_batch VALUES ($1, $2, $3)");
            foreach (var field in fields)
            {
                command.Parameters.Add(new() { Value = field });
            }
            batch.BatchCommands.Add(command);
        }
        try
        {
            await batch.ExecuteNonQueryAsync();
        }
        catch (PostgresException e)
        {
            Console.WriteLine($"{e.SqlState}: {e.MessageText}");
            Console.WriteLine(e.Detail);
        }
        await using var count = new NpgsqlCommand("SELECT count(*) FROM ga.package_refs_batch", connection);
        Console.WriteLine($"rows in the table: {await count.ExecuteScalarAsync()}");
    }

    // Not compared by check.sh: the timings depend on the machine. Two rounds: the first one includes the JIT.
    public static async Task Timings()
    {
        var rows = PackageRefs().ToList();
        await using var dataSource = Db.DataSource();
        await using var connection = await dataSource.OpenConnectionAsync();

        async Task Reset()
        {
            await using var command = new NpgsqlCommand("TRUNCATE ga.package_refs_import", connection);
            await command.ExecuteNonQueryAsync();
        }

        for (var round = 1; round <= 2; round++)
        {
            await Reset();
            var watch = Stopwatch.StartNew();
            foreach (var fields in rows)
            {
                await using var command = new NpgsqlCommand("INSERT INTO ga.package_refs_import VALUES ($1, $2, $3)", connection);
                foreach (var field in fields)
                {
                    command.Parameters.Add(new() { Value = field });
                }
                await command.ExecuteNonQueryAsync();
            }
            Console.WriteLine($"round {round}: {rows.Count} INSERT commands: {watch.ElapsedMilliseconds} ms");

            await Reset();
            watch.Restart();
            await using (var batch = new NpgsqlBatch(connection))
            {
                foreach (var fields in rows)
                {
                    var command = new NpgsqlBatchCommand("INSERT INTO ga.package_refs_import VALUES ($1, $2, $3)");
                    foreach (var field in fields)
                    {
                        command.Parameters.Add(new() { Value = field });
                    }
                    batch.BatchCommands.Add(command);
                }
                await batch.ExecuteNonQueryAsync();
            }
            Console.WriteLine($"round {round}: one NpgsqlBatch of {rows.Count} INSERTs: {watch.ElapsedMilliseconds} ms");

            await Reset();
            watch.Restart();
            await using (var writer = await connection.BeginBinaryImportAsync(
                "COPY ga.package_refs_import (project, package, version) FROM STDIN (FORMAT binary)"))
            {
                foreach (var fields in rows)
                {
                    await writer.WriteRowAsync(default, fields[0], fields[1], fields[2]);
                }
                await writer.CompleteAsync();
            }
            Console.WriteLine($"round {round}: binary COPY of {rows.Count} rows: {watch.ElapsedMilliseconds} ms");
        }
    }

    public static async Task TooMany()
    {
        // No pool: each Open is a new server process, until max_connections (100 by default) is reached
        await using var dataSource = Db.DataSource("Pooling=false");
        var connections = new List<NpgsqlConnection>();
        try
        {
            for (var i = 0; i < 200; i++)
            {
                connections.Add(await dataSource.OpenConnectionAsync());
            }
        }
        catch (PostgresException e)
        {
            Console.WriteLine($"{e.SqlState}: {e.MessageText}");
        }
        finally
        {
            foreach (var connection in connections)
            {
                await connection.DisposeAsync();
            }
        }
    }
}

File: code/postgresql-aurora/java/src/main/java/dev/learn/pg/L04.java
package dev.learn.pg;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import java.io.BufferedReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.BatchUpdateException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.postgresql.PGConnection;
import org.postgresql.util.PSQLException;

/** Lesson 4: PostgreSQL from Java with the PostgreSQL JDBC driver, on the course model (sql/schema.sql). */
final class L04 {
    private L04() {
    }

    // The server of server.sh; the course runs as postgres to keep the examples short (lesson 1 creates roles)
    static String url() {
        String url = System.getenv("PG_JDBC_URL");
        return url != null ? url
                : "jdbc:postgresql://localhost:5432/learn?user=postgres&password=learn&ApplicationName=learn-java";
    }

    static void connect() throws SQLException {
        try (Connection connection = DriverManager.getConnection(url());
             Statement statement = connection.createStatement();
             ResultSet rs = statement.executeQuery(
                     "SELECT current_user, current_setting('server_version'), (SELECT count(*) FROM ci.runs),"
                             + " current_setting('TimeZone')")) {
            rs.next();
            System.out.println("driver: " + connection.getMetaData().getDriverVersion());
            // The driver sets the session's TimeZone to the JVM's default time zone
            System.out.println("JVM time zone: " + java.util.TimeZone.getDefault().getID() + ", session TimeZone: " + rs.getString(4));
            System.out.println("user: " + rs.getString(1));
            System.out.println("server: " + rs.getString(2));
            System.out.println("runs: " + rs.getLong(3));
        }
    }

    static void parameters() throws SQLException {
        try (Connection connection = DriverManager.getConnection(url())) {
            // JDBC's ? placeholders; the driver sends them as $1, $2
            try (PreparedStatement ps = connection.prepareStatement("""
                    SELECT run_id, started_at FROM ci.runs
                    WHERE workflow_name = ? AND started_at >= ?
                    ORDER BY started_at, run_id
                    LIMIT 2
                    """)) {
                ps.setString(1, "Rust course examples");
                ps.setObject(2, OffsetDateTime.of(2026, 9, 13, 18, 0, 0, 0, ZoneOffset.UTC));
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        System.out.println(rs.getLong(1) + " " + rs.getObject(2, OffsetDateTime.class));
                    }
                }
            }

            // An array parameter: = ANY (?)
            try (PreparedStatement ps = connection.prepareStatement(
                    "SELECT labels[1], count(*) FROM ci.jobs WHERE labels[1] = ANY (?) GROUP BY 1 ORDER BY 1")) {
                ps.setArray(1, connection.createArrayOf("text", new String[] {"macos-latest", "windows-latest"}));
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        System.out.println(rs.getString(1) + ": " + rs.getLong(2));
                    }
                }
            }

            // An OffsetDateTime keeps its offset; the server compares instants
            try (PreparedStatement ps = connection.prepareStatement("SELECT count(*) FROM ci.runs WHERE started_at >= ?")) {
                ps.setObject(1, OffsetDateTime.of(2026, 9, 14, 0, 0, 0, 0, ZoneOffset.ofHours(-4)));
                try (ResultSet rs = ps.executeQuery()) {
                    rs.next();
                    System.out.println("runs since 2026-09-14 00:00 in Toronto: " + rs.getLong(1));
                }
            }

            // setString binds a varchar: PostgreSQL has no bigint = varchar operator
            try (PreparedStatement ps = connection.prepareStatement("SELECT count(*) FROM ci.runs WHERE run_id = ?")) {
                ps.setString(1, "34852867099");
                ps.executeQuery();
            } catch (PSQLException e) {
                System.out.println(e.getSQLState() + ": " + e.getServerErrorMessage().getMessage());
            }
        }
    }

    static int backendPid(Connection connection) throws SQLException {
        try (Statement statement = connection.createStatement();
             ResultSet rs = statement.executeQuery("SELECT pg_backend_pid()")) {
            rs.next();
            return rs.getInt(1);
        }
    }

    static void pool() throws SQLException {
        Set<Integer> backends = new HashSet<>();
        for (int i = 0; i < 20; i++) {
            try (Connection connection = DriverManager.getConnection(url())) {
                backends.add(backendPid(connection));
            }
        }
        System.out.println("DriverManager: 20 connections, " + backends.size() + " server process(es)");

        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(url().replace("learn-java", "learn-hikari"));
        config.setMaximumPoolSize(5);
        try (HikariDataSource pool = new HikariDataSource(config)) {
            backends.clear();
            for (int i = 0; i < 20; i++) {
                try (Connection connection = pool.getConnection()) {
                    backends.add(backendPid(connection));
                }
            }
            System.out.println("HikariCP: 20 connections, " + backends.size() + " server process(es)");
            // HikariCP opens connections in the background until minimumIdle, which defaults to maximumPoolSize
            try (Connection connection = pool.getConnection()) {
                long deadline = System.currentTimeMillis() + 10_000;
                long sessions;
                do {
                    try (Statement statement = connection.createStatement();
                         ResultSet rs = statement.executeQuery(
                                 "SELECT count(*) FROM pg_stat_activity WHERE application_name = 'learn-hikari'")) {
                        rs.next();
                        sessions = rs.getLong(1);
                    }
                } while (sessions < 5 && System.currentTimeMillis() < deadline && sleep(100));
                System.out.println("sessions opened by the pool: " + sessions);
            }
        }
    }

    private static boolean sleep(long millis) {
        try {
            Thread.sleep(millis);
            return true;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return false;
        }
    }

    static void prepare() throws SQLException {
        // The driver switches to a named server-side prepared statement on the prepareThreshold-th execution (default 5)
        try (Connection connection = DriverManager.getConnection(url());
             PreparedStatement count = connection.prepareStatement(
                     "SELECT count(*) FROM pg_prepared_statements WHERE statement ~ 'ci\\.jobs WHERE'")) {
            count.unwrap(org.postgresql.PGStatement.class).setPrepareThreshold(0);
            for (int i = 1; i <= 6; i++) {
                try (PreparedStatement ps = connection.prepareStatement("SELECT count(*) FROM ci.jobs WHERE run_id = ?")) {
                    ps.setLong(1, 34852867099L);
                    ps.executeQuery().close();
                }
                try (ResultSet rs = count.executeQuery()) {
                    rs.next();
                    System.out.println("after " + i + " execution(s): " + rs.getLong(1) + " prepared statement(s)");
                }
            }
        }
    }

    static void copy() throws Exception {
        try (Connection connection = DriverManager.getConnection(url())) {
            try (Statement statement = connection.createStatement()) {
                statement.execute("""
                        DROP TABLE IF EXISTS ga.package_refs_import;
                        CREATE TABLE ga.package_refs_import (LIKE ga.package_refs INCLUDING ALL);
                        """);
            }
            // COPY ... FROM STDIN: the file is on the client, the server parses the CSV
            try (BufferedReader csv = Files.newBufferedReader(Path.of("../ladybugdb/data/ga/package_refs.csv"))) {
                long rows = connection.unwrap(PGConnection.class).getCopyAPI()
                        .copyIn("COPY ga.package_refs_import FROM STDIN (FORMAT csv, HEADER match)", csv);
                System.out.println("copied: " + rows + " rows");
            }
            try (Statement statement = connection.createStatement();
                 ResultSet rs = statement.executeQuery("""
                         SELECT (SELECT count(*) FROM (TABLE ga.package_refs EXCEPT TABLE ga.package_refs_import) AS d),
                                (SELECT count(*) FROM (TABLE ga.package_refs_import EXCEPT TABLE ga.package_refs) AS d)
                         """)) {
                rs.next();
                System.out.println("rows only in the server-side load: " + rs.getLong(1)
                        + ", only in the client copy: " + rs.getLong(2));
            }
        }
    }

    static void generatedKeys() throws SQLException {
        try (Connection connection = DriverManager.getConnection(url())) {
            try (Statement statement = connection.createStatement()) {
                statement.execute("""
                        DROP TABLE IF EXISTS ci.notes;
                        CREATE TABLE ci.notes (
                            note_id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
                            run_id  bigint NOT NULL REFERENCES ci.runs,
                            body    text NOT NULL,
                            created timestamptz NOT NULL DEFAULT '2026-09-15 00:00:00+00'
                        );
                        """);
            }
            // RETURN_GENERATED_KEYS: the driver appends RETURNING * to the INSERT
            try (PreparedStatement ps = connection.prepareStatement(
                    "INSERT INTO ci.notes (run_id, body) VALUES (?, ?)", Statement.RETURN_GENERATED_KEYS)) {
                ps.setLong(1, 34852867099L);
                ps.setString(2, "first deployment of the day");
                ps.executeUpdate();
                try (ResultSet keys = ps.getGeneratedKeys()) {
                    keys.next();
                    int columns = keys.getMetaData().getColumnCount();
                    StringBuilder line = new StringBuilder("generated keys:");
                    for (int c = 1; c <= columns; c++) {
                        line.append(' ').append(keys.getMetaData().getColumnName(c)).append('=').append(keys.getString(c));
                    }
                    System.out.println(line);
                }
            }
            // Naming the columns asks for RETURNING note_id only
            try (PreparedStatement ps = connection.prepareStatement(
                    "INSERT INTO ci.notes (run_id, body) VALUES (?, ?)", new String[] {"note_id"})) {
                ps.setLong(1, 34852867099L);
                ps.setString(2, "second note");
                ps.executeUpdate();
                try (ResultSet keys = ps.getGeneratedKeys()) {
                    keys.next();
                    System.out.println("generated keys: " + keys.getMetaData().getColumnCount() + " column, note_id="
                            + keys.getLong("note_id"));
                }
            }
        }
    }

    static void exerciseBatch() throws Exception {
        try (Connection connection = DriverManager.getConnection(url() + "&reWriteBatchedInserts=true")) {
            try (Statement statement = connection.createStatement()) {
                statement.execute("""
                        DROP TABLE IF EXISTS ga.package_refs_batch;
                        CREATE TABLE ga.package_refs_batch (LIKE ga.package_refs INCLUDING ALL);
                        """);
            }
            List<String> lines = Files.readAllLines(Path.of("../ladybugdb/data/ga/package_refs.csv"));
            List<String> rows = new java.util.ArrayList<>(lines.subList(1, lines.size()));
            rows.add(rows.get(0));
            connection.setAutoCommit(false);
            try (PreparedStatement ps = connection.prepareStatement("INSERT INTO ga.package_refs_batch VALUES (?, ?, ?)")) {
                for (String row : rows) {
                    String[] fields = row.split(",");
                    ps.setString(1, fields[0]);
                    ps.setString(2, fields[1]);
                    ps.setString(3, fields[2]);
                    ps.addBatch();
                }
                ps.executeBatch();
                connection.commit();
            } catch (BatchUpdateException e) {
                connection.rollback();
                SQLException cause = e.getNextException();
                System.out.println(cause.getSQLState() + ": " + ((PSQLException) cause).getServerErrorMessage().getMessage());
            }
            connection.setAutoCommit(true);
            try (Statement statement = connection.createStatement();
                 ResultSet rs = statement.executeQuery("SELECT count(*) FROM ga.package_refs_batch")) {
                rs.next();
                System.out.println("rows in the table: " + rs.getLong(1));
            }
        }
    }

    static void target() {
        // targetServerType: before handing out a connection, the driver checks the server's role, as it would for each
        // host of an Aurora cluster; the local server is a primary
        for (String target : List.of("primary", "preferSecondary", "secondary")) {
            try (Connection connection = DriverManager.getConnection(url() + "&targetServerType=" + target);
                 Statement statement = connection.createStatement();
                 ResultSet rs = statement.executeQuery("SELECT pg_is_in_recovery()")) {
                rs.next();
                System.out.println(target + ": connected, pg_is_in_recovery() = " + rs.getBoolean(1));
            } catch (SQLException e) {
                System.out.println(target + ": " + e.getSQLState() + ": " + e.getMessage());
            }
        }
    }
}
